### PR TITLE
Add initial passengers functionality.

### PR DESCRIPTION
**Pull request by GenKyoko**
[SOMEVEHICLE]
InitPassengers=<List of InfantryTypes/UnitTypes>
InitPassengersNum=<A list of the number of objects>

### CODE REVIEW

File: src/extensions/foot/footext_hooks.cpp [hunk 1]
```diff
@@ -35,9 +35,94 @@
...
+#include "tibsun_defines.h"
+
+/*
+* The Foot object initializes the passenger.
+* 
+* Injection objective function: FootClass::Unlimbo
+* 
+* @author: RossinCarlinx
+*/
+static void __stdcall FootClass_Initialize_Passengers(FootClass* this_ptr, TechnoTypeClassExtension* technotypeext) {
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:52]: 1. This function can be moved to FootClassExtension as a member function.
2. The param of TechnoTypeClassExtension is not required as you can perform the extension type class fetch starting from `this_ptr->Techno_Type_Class()`

File: src/extensions/foot/footext_hooks.cpp [hunk 2]
```diff
@@ -35,9 +35,94 @@
...
+    auto const Count = technotypeext->InitPassengers.Count();
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:53]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 3]
```diff
@@ -35,9 +35,94 @@
...
+    // Check if the initial number of passenger types is greater than 0.
+    if (Count > 0) {
+
+        auto const house = this_ptr->House;
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:57]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 4]
```diff
@@ -35,9 +35,94 @@
...
+        // Iterates through objects in InitPassenger, adding them if they are infantry or units.
+        for (int i = 0;
+            i < Count;
+            i++) {
+            auto const technotype = technotypeext->InitPassengers[i];
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:62]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 5]
```diff
@@ -35,9 +35,94 @@
...
+            auto const rtti = technotype->Kind_Of();
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:63]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 6]
```diff
@@ -35,9 +35,94 @@
...
+
+            if (rtti != RTTI_INFANTRYTYPE &&
+                rtti != RTTI_UNITTYPE) {
+                continue;
+            }
+
+            auto num = (technotypeext->InitPassengerNums.Count() > i) ?
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:70]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 7]
```diff
@@ -35,9 +35,94 @@
...
+                technotypeext->InitPassengerNums[i] :
+                1;
+
+            // A loop that creates objects and puts them into the position of passengers.
+            do {
+                auto const object = technotype->Create_One_Of(house);
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:76]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 8]
```diff
@@ -35,9 +35,94 @@
...
+                auto const passenger = object->As_Foot();
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:77]: Please do not use `auto`, use explicit types.

File: src/extensions/foot/footext_hooks.cpp [hunk 9]
```diff
@@ -35,9 +35,94 @@
...
+                // Prevent the generation of initial occupants indefinitely.
+                if (passenger->Class_Of() == this_ptr->Class_Of()) {
+                    Extension::Fetch<TechnoClassExtension>(passenger)->Initialized = true;
+                }
+
+                passenger->Set_Coord(this_ptr->Get_Coord());
+                passenger->Limbo();
+                this_ptr->Cargo.Attach(passenger);
+#ifndef NDEBUG
+                Vinifera_Output_Debug_String("this_ptr->Cargo.Attach();");
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:91]: Use `DEBUG_INFO` (or similar) macro here for log printing.

File: src/extensions/foot/footext_hooks.cpp [hunk 10]
```diff
@@ -35,9 +35,94 @@
...
+
+    // If it has already been initialized, it is automatically skipped.
+    if (technoclassext->Initialized) {
+        goto normal;
+    }
+
+    technoclassext->Initialized = true;
 
+    // Perform initialization of passengers.
+    FootClass_Initialize_Passengers(this_ptr, technotypeext);
```
Comment (CCHyper) [inline src/extensions/foot/footext_hooks.cpp:118]: See the first comment of this review. You can perform a extension class lookup on `this_ptr` and the function can be moved to FootClassExtension.

File: src/extensions/techno/technoext.h [hunk 11]
```diff
@@ -73,4 +73,13 @@ class TechnoClassExtension : public RadioClassExtension
          *  The current electric bolt instance fired by this object.
          */
         EBoltClass *ElectricBolt;
+
+        /**
+         * Is the object appearing for the first time?
+         */
+        bool Initialized;
```
Comment (CCHyper) [inline src/extensions/techno/technoext.h:80]: Please use `Is...` or `Has...` prefixes for boolean property names.

File: src/extensions/technotype/technotypeext.cpp [hunk 12]
```diff
@@ -112,6 +114,13 @@ HRESULT TechnoTypeClassExtension::Load(IStream *pStm)
         return E_FAIL;
     }
 
+    InitPassengers.Clear();
+    InitPassengerNums.Clear();
+
```
Comment (CCHyper) [inline src/extensions/technotype/technotypeext.cpp:119]: Double white space here.

File: src/extensions/technotype/technotypeext.h [hunk 13]
```diff
@@ -158,4 +158,14 @@ class TechnoTypeClassExtension : public ObjectTypeClassExtension
...
+
+        /**
+         * List of initial passengers.
+         */
+        TypeList<int> InitPassengerNums;
+
+        /**
+         * Initial list of passenger types.
+         */
+        TypeList<TechnoTypeClass*> InitPassengers;
```
Comment (CCHyper) [inline src/extensions/technotype/technotypeext.h:170]: The TypeList type here can be `const` as the types are not modified.

**Review by CCHyper — CHANGES_REQUESTED** [review]

**Comment by CCHyper** [comment]
@GenKyoko I have submitted a review for you. Could you please provide a detail PR body description for this PR and the usage behind it, you can see other PRs for an example of the desired information. Thanks.

File: src/extensions/technotype/technotypeext.h (hunk 13 above)
Comment (GenKyoko) [reply src/extensions/technotype/technotypeext.h:170]: I've tested this now and it seems to cause an error if you use const. I'm sorry I didn't have time to check the features during these two months.

Comment (CCHyper) [reply src/extensions/technotype/technotypeext.h:170]: Did you try `TypeList<const TechnoTypeClass*> InitPassengers;` ?

Comment (GenKyoko) [reply src/extensions/technotype/technotypeext.h:170]: Yes.

Comment (GenKyoko) [reply src/extensions/technotype/technotypeext.h:170]: 屏幕截图 2024-02-18 234659

File: src/extensions/foot/footext_hooks.cpp (hunk 1 above)
Comment (GenKyoko) [reply src/extensions/foot/footext_hooks.cpp]: This has been moved to the FootExt.cpp and the parameters have been removed, do you want to merge it?

**Pull request closed without merging**